File: docker/python/trace_collector.py
"""
FlowMind Python trace collector — Section 7.6
Injected into Docker container at /opt/flowmind/trace_collector.py
Imported by instrumented user code as: from trace_collector import trace
"""
import json, sys, atexit, builtins

MAX_STEPS = 10000
MAX_DEPTH = 500
MAX_ARRAY_LENGTH = 100
MAX_OBJECT_KEYS = 50
MAX_STRING_LENGTH = 1000

_steps = []
_step_index = 0
_frame_stack = [{"id": "frame_0", "name": "<module>", "depth": 0, "vars": {}, "parentId": None}]
_frame_counter = 1
_truncated = False


def _serialize(value, depth=0):
    if depth > 5:
        return {"type": "object", "value": "[Max Depth]"}
    if value is None:
        return {"type": "none", "value": None}
    if isinstance(value, bool):
        return {"type": "bool", "value": value}
    if isinstance(value, int):
        return {"type": "int", "value": value}
    if isinstance(value, float):
        return {"type": "float", "value": value}
    if isinstance(value, str):
        return {"type": "str", "value": value[:MAX_STRING_LENGTH]}
    if isinstance(value, (list, tuple)):
        kind = "list" if isinstance(value, list) else "tuple"
        return {
            "type": kind,
            "value": [_serialize(v, depth + 1) for v in value[:MAX_ARRAY_LENGTH]],
            "length": len(value),
        }
    if isinstance(value, dict):
        items = list(value.items())[:MAX_OBJECT_KEYS]
        return {
            "type": "dict",
            "value": {str(k): _serialize(v, depth + 1) for k, v in items},
        }
    if callable(value):
        return {"type": "function", "value": f"[function {getattr(value, '__name__', '?')}]"}
    return {"type": "object", "value": str(value)[:MAX_STRING_LENGTH]}


def _get_frame():
    return _frame_stack[-1]


def _get_vars():
    return {k: _serialize(v) for k, v in _get_frame()["vars"].items()}


def _push(step):
    global _truncated
    if len(_steps) >= MAX_STEPS:
        _truncated = True
        return
    _steps.append(step)


class _Trace:
    def line(self, line_num):
        global _step_index
        _push({"index": _step_index, "type": "line", "line": line_num,
               "frameId": _get_frame()["id"], "variables": _get_vars(), "timestamp": _step_index})
        _step_index += 1

    def call(self, name, **kwargs):
        global _step_index, _frame_counter
        if len(_frame_stack) >= MAX_DEPTH:
            return
        parent = _get_frame()
        new_frame = {
            "id": f"frame_{_frame_counter}",
            "name": name,
            "depth": parent["depth"] + 1,
            "vars": dict(kwargs),
            "parentId": parent["id"],
        }
        _frame_counter += 1
        _frame_stack.append(new_frame)
        _push({
            "index": _step_index, "type": "call", "functionName": name,
            "arguments": {k: _serialize(v) for k, v in kwargs.items()},
            "frameId": new_frame["id"], "parentFrameId": parent["id"],
            "depth": new_frame["depth"], "timestamp": _step_index,
        })
        _step_index += 1

    def return_(self, value):
        global _step_index
        frame = _frame_stack.pop() if len(_frame_stack) > 1 else _frame_stack[0]
        _push({"index": _step_index, "type": "return", "functionName": frame["name"],
               "returnValue": _serialize(value), "frameId": frame["id"],
               "depth": frame["depth"], "timestamp": _step_index})
        _step_index += 1
        return value

    def assign(self, name, old_val, new_val):
        global _step_index
        _get_frame()["vars"][name] = new_val
        _push({"index": _step_index, "type": "assign", "variable": name,
               "oldValue": _serialize(old_val), "newValue": _serialize(new_val),
               "frameId": _get_frame()["id"], "timestamp": _step_index})
        _step_index += 1
        return new_val

    def condition(self, expr, result, branch=None):
        global _step_index
        _push({"index": _step_index, "type": "condition", "expression": str(expr),
               "result": bool(result), "branch": "true" if result else "false",
               "frameId": _get_frame()["id"], "timestamp": _step_index})
        _step_index += 1

    def loop(self, line_num, **vars_snapshot):
        global _step_index
        _push({"index": _step_index, "type": "loop_iter", "loopLine": line_num,
               "variables": {k: _serialize(v) for k, v in vars_snapshot.items()},
               "frameId": _get_frame()["id"], "timestamp": _step_index})
        _step_index += 1

    def output(self, text, stream="stdout"):
        global _step_index
        _push({"index": _step_index, "type": "output", "text": str(text), "stream": stream,
               "frameId": _get_frame()["id"], "timestamp": _step_index})
        _step_index += 1

    def exception(self, error_type, message, stack_trace=""):
        global _step_index
        _push({"index": _step_index, "type": "exception", "errorType": error_type,
               "message": message, "stackTrace": stack_trace,
               "frameId": _get_frame()["id"], "timestamp": _step_index})
        _step_index += 1

    def get_trace(self):
        return {"steps": _steps, "frames": _frame_stack, "truncated": _truncated}


trace = _Trace()


# Intercept print()
_original_print = builtins.print

def _patched_print(*args, sep=" ", end="\n", file=None, flush=False):
    text = sep.join(str(a) for a in args)
    stream = "stderr" if file is sys.stderr else "stdout"
    trace.output(text + (end if end != "\n" else ""), stream)
    _original_print(*args, sep=sep, end=end, file=file, flush=flush)

builtins.print = _patched_print


# Write trace on exit
def _on_exit():
    try:
        with open("/tmp/trace.json", "w") as f:
            json.dump(trace.get_trace(), f)
    except Exception:
        pass

atexit.register(_on_exit)


# Intercept uncaught exceptions
import traceback as _traceback
_original_excepthook = sys.excepthook

def _excepthook(exc_type, exc_value, exc_tb):
    trace.exception(exc_type.__name__, str(exc_value), "".join(_traceback.format_tb(exc_tb)))
    _on_exit()
    _original_excepthook(exc_type, exc_value, exc_tb)

sys.excepthook = _excepthook

File: docker/node/trace-collector.js
/**
 * FlowMind Node.js trace collector — Section 7.5
 * Injected into Docker container at /opt/flowmind/trace-collector.js
 * Required by instrumented user code as: const __trace = require('/opt/flowmind/trace-collector');
 */
'use strict';

const MAX_STEPS = 10000;
const MAX_DEPTH = 500;
const MAX_ARRAY_LENGTH = 100;
const MAX_OBJECT_KEYS = 50;
const MAX_STRING_LENGTH = 1000;

const steps = [];
let stepIndex = 0;
const frameStack = [{ id: 'frame_0', name: '<global>', depth: 0, vars: {}, parentId: null }];
let frameCounter = 1;
let truncated = false;

function serialize(value, depth) {
  if (depth === undefined) depth = 0;
  if (depth > 5) return { type: 'object', value: '[Max Depth]' };
  if (value === null) return { type: 'null', value: null };
  if (value === undefined) return { type: 'undefined', value: undefined };
  const t = typeof value;
  if (t === 'number' || t === 'boolean') return { type: t, value: value };
  if (t === 'string') return { type: 'string', value: value.slice(0, MAX_STRING_LENGTH) };
  if (Array.isArray(value)) {
    return {
      type: 'array',
      value: value.slice(0, MAX_ARRAY_LENGTH).map(function(v) { return serialize(v, depth + 1); }),
      length: value.length,
    };
  }
  if (t === 'function') return { type: 'function', value: '[Function: ' + (value.name || 'anonymous') + ']' };
  if (t === 'object') {
    const entries = Object.entries(value).slice(0, MAX_OBJECT_KEYS);
    const obj = {};
    for (const pair of entries) {
      obj[pair[0]] = serialize(pair[1], depth + 1);
    }
    return { type: 'object', value: obj };
  }
  return { type: 'unknown', value: String(value) };
}

function getCurrentFrame() {
  return frameStack[frameStack.length - 1];
}

function getVariables() {
  const frame = getCurrentFrame();
  const vars = {};
  const entries = Object.entries(frame.vars);
  for (const pair of entries) {
    vars[pair[0]] = serialize(pair[1]);
  }
  return vars;
}

function push(step) {
  if (steps.length >= MAX_STEPS) {
    truncated = true;
    return;
  }
  steps.push(step);
}

const __trace = {
  line: function(lineNum) {
    push({ index: stepIndex++, type: 'line', line: lineNum, frameId: getCurrentFrame().id, variables: getVariables(), timestamp: stepIndex });
  },
  call: function(name, args) {
    if (frameStack.length >= MAX_DEPTH) return;
    const parentFrame = getCurrentFrame();
    const newFrame = {
      id: 'frame_' + (frameCounter++),
      name: name,
      depth: parentFrame.depth + 1,
      vars: Object.assign({}, args || {}),
      parentId: parentFrame.id,
    };
    frameStack.push(newFrame);
    const serializedArgs = {};
    const argEntries = Object.entries(args || {});
    for (const pair of argEntries) {
      serializedArgs[pair[0]] = serialize(pair[1]);
    }
    push({ index: stepIndex++, type: 'call', functionName: name, arguments: serializedArgs, frameId: newFrame.id, parentFrameId: parentFrame.id, depth: newFrame.depth, timestamp: stepIndex });
  },
  return: function(value) {
    const frame = frameStack.length > 1 ? frameStack.pop() : frameStack[0];
    push({ index: stepIndex++, type: 'return', functionName: frame.name, returnValue: serialize(value), frameId: frame.id, depth: frame.depth, timestamp: stepIndex });
    return value;
  },
  assign: function(name, oldVal, newVal) {
    getCurrentFrame().vars[name] = newVal;
    push({ index: stepIndex++, type: 'assign', variable: name, oldValue: serialize(oldVal), newValue: serialize(newVal), frameId: getCurrentFrame().id, timestamp: stepIndex });
    return newVal;
  },
  condition: function(expr, result, branch) {
    push({ index: stepIndex++, type: 'condition', expression: expr, result: !!result, branch: result ? 'true' : 'false', frameId: getCurrentFrame().id, timestamp: stepIndex });
  },
  loop: function(line, vars) {
    const serializedVars = {};
    if (vars) {
      const entries = Object.entries(vars);
      for (const pair of entries) {
        serializedVars[pair[0]] = serialize(pair[1]);
      }
    }
    push({ index: stepIndex++, type: 'loop_iter', loopLine: line, variables: serializedVars, frameId: getCurrentFrame().id, timestamp: stepIndex });
  },
  output: function(text, stream) {
    push({ index: stepIndex++, type: 'output', text: String(text), stream: stream || 'stdout', frameId: getCurrentFrame().id, timestamp: stepIndex });
  },
  exception: function(errorType, message, stackTrace) {
    push({ index: stepIndex++, type: 'exception', errorType: errorType, message: message, stackTrace: stackTrace, frameId: getCurrentFrame().id, timestamp: stepIndex });
  },
  getTrace: function() {
    return { steps: steps, frames: frameStack, truncated: truncated };
  },
};

// Intercept console.log/warn/error
const originalLog = console.log;
const originalWarn = console.warn;
const originalError = console.error;

console.log = function() {
  const text = Array.from(arguments).map(function(a) { return typeof a === 'string' ? a : JSON.stringify(a); }).join(' ');
  __trace.output(text, 'stdout');
  originalLog.apply(console, arguments);
};
console.warn = function() {
  const text = Array.from(arguments).map(function(a) { return typeof a === 'string' ? a : JSON.stringify(a); }).join(' ');
  __trace.output(text, 'stderr');
  originalWarn.apply(console, arguments);
};
console.error = function() {
  const text = Array.from(arguments).map(function(a) { return typeof a === 'string' ? a : JSON.stringify(a); }).join(' ');
  __trace.output(text, 'stderr');
  originalError.apply(console, arguments);
};

// Write trace.json on process exit
process.on('exit', function() {
  try {
    require('fs').writeFileSync('/tmp/trace.json', JSON.stringify(__trace.getTrace()));
  } catch(e) { /* ignore */ }
});

process.on('uncaughtException', function(err) {
  __trace.exception(err.constructor.name, err.message, err.stack);
  try {
    require('fs').writeFileSync('/tmp/trace.json', JSON.stringify(__trace.getTrace()));
  } catch(e) { /* ignore */ }
  process.exit(1);
});

module.exports = __trace;
